{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List datasets that will be used for training with scBasset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# maybe anndata and others required\n",
    "# !pip install anndata h5py\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.8.0'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import h5py\n",
    "h5py.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.8.0'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import anndata\n",
    "anndata.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-06-17 01:42:24.913070: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX_VNNI FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-06-17 01:42:24.967882: I tensorflow/core/util/port.cc:104] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2024-06-17 01:42:24.969910: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2024-06-17 01:42:24.969919: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2024-06-17 01:42:25.295732: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory\n",
      "2024-06-17 01:42:25.295775: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory\n",
      "2024-06-17 01:42:25.295778: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "import scbasset\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# d = '/mnt/c/Users/IgnacioIbarra/Dropbox/workspace/theislab/mubind-pipeline/notebooks/pipeline/01_train_n_predict'\n",
    "d = '/home/ilibarra/workspace/theislab/mubind-pipeline/notebooks/pipeline/01_train_n_predict'\n",
    "os.chdir(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ilibarra/workspace/theislab/mubind-pipeline/notebooks/pipeline/01_train_n_predict'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import utils\n",
    "import glob\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the datasets used in the core benchmark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'organoids': '/home/ilibarra/workspace/theislab/mubind/data/organoids_treutlein_dataset/RNA_ATAC_metacells_sce_peaks_obs2000_var4000.h5ad',\n",
       " 'gbm': '/home/ilibarra/workspace/theislab/mubind/data/gbm_multiome_scdori/23_01_23_atac_compressed_n1000.h5ad',\n",
       " 'noack_2022': '/home/ilibarra/workspace/theislab/mubind/data/noack_2022/*/merged_scATAC_integrated_cicero_faye_chong_obs*_var*.h5ad',\n",
       " 'pancreatic_endocrinogenesis': '/home/ilibarra/workspace/theislab/mubind/data/pancreatic_endocrinogenesis/*/pancreas_multiome_2022_processed_atac_obs*_var*.h5ad',\n",
       " 'pbmc': '/home/ilibarra/workspace/theislab/mubind/data/pbmc/*/processed_laura_2023_obs*_var*.h5ad'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# data_directory = '/mnt/f/workspace/theislab/mubind/data'\n",
    "data_directory = '/home/ilibarra/workspace/theislab/mubind/data'\n",
    "\n",
    "path_by_dataset = utils.get_datasets(data_directory=data_directory)\n",
    "path_by_dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# os.chdir('/mnt/c/Users/IgnacioIbarra/Dropbox/workspace/scbasset/bin')\n",
    "os.chdir('/home/ilibarra/workspace/scBasset/bin')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run python using the scbasset environment. If not found, please set up manually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pythonbin = '/home/ilibarra/.conda/envs/scbasset/bin/python' # ''\n",
    "pythonbin = '/home/ilibarra/miniconda3/envs/scbasset/bin/python'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata = anndata.read_h5ad(ad_path)\n",
    "# print(adata.shape)\n",
    "# adata.var['chr'] = adata.var_names.str.split('-').str[0]\n",
    "# adata.var['chr'] = np.where(~adata.var['chr'].str.contains('chr'), 'chr', '') + adata.var['chr']\n",
    "# adata.var['start'] = adata.var_names.str.split('-').str[1]\n",
    "# adata.var['end'] = adata.var_names.str.split('-').str[2]\n",
    "# adata.write(ad_path, compression='lzf')\n",
    "# genome_fasta = '/mnt/c/Users/ignacio.ibarra/Dropbox/annotations/%s/genome/%s.fa' % (species, species)\n",
    "# os.path.exists(genome_fasta)\n",
    "# print(genome_fasta)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "organoids /home/ilibarra/workspace/theislab/mubind/data/organoids_treutlein_dataset/RNA_ATAC_metacells_sce_peaks_obs2000_var4000.h5ad\n",
      "gbm /home/ilibarra/workspace/theislab/mubind/data/gbm_multiome_scdori/23_01_23_atac_compressed_n1000.h5ad\n",
      "noack_2022 /home/ilibarra/workspace/theislab/mubind/data/noack_2022/*/merged_scATAC_integrated_cicero_faye_chong_obs*_var*.h5ad\n",
      "pancreatic_endocrinogenesis /home/ilibarra/workspace/theislab/mubind/data/pancreatic_endocrinogenesis/*/pancreas_multiome_2022_processed_atac_obs*_var*.h5ad\n",
      "pbmc /home/ilibarra/workspace/theislab/mubind/data/pbmc/*/processed_laura_2023_obs*_var*.h5ad\n"
     ]
    }
   ],
   "source": [
    "for k in path_by_dataset:\n",
    "    print(k, path_by_dataset[k])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_paths = [[k, ad_path] for k in path_by_dataset for ad_path in glob.glob(path_by_dataset[k])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>path</th>\n",
       "      <th>path_exists</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>10000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>organoids</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>noack_2022</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>pancreatic_endocrinogenesis</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>pbmc</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gbm</td>\n",
       "      <td>/home/ilibarra/workspace/theislab/mubind/data/...</td>\n",
       "      <td>True</td>\n",
       "      <td>compressed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        dataset  \\\n",
       "19  pancreatic_endocrinogenesis   \n",
       "28  pancreatic_endocrinogenesis   \n",
       "36                         pbmc   \n",
       "43                         pbmc   \n",
       "6                    noack_2022   \n",
       "11                   noack_2022   \n",
       "16  pancreatic_endocrinogenesis   \n",
       "25  pancreatic_endocrinogenesis   \n",
       "32                         pbmc   \n",
       "39                         pbmc   \n",
       "2                    noack_2022   \n",
       "8                    noack_2022   \n",
       "15  pancreatic_endocrinogenesis   \n",
       "24  pancreatic_endocrinogenesis   \n",
       "37                         pbmc   \n",
       "44                         pbmc   \n",
       "20  pancreatic_endocrinogenesis   \n",
       "29  pancreatic_endocrinogenesis   \n",
       "3                    noack_2022   \n",
       "9                    noack_2022   \n",
       "0                     organoids   \n",
       "13  pancreatic_endocrinogenesis   \n",
       "22  pancreatic_endocrinogenesis   \n",
       "35                         pbmc   \n",
       "42                         pbmc   \n",
       "21  pancreatic_endocrinogenesis   \n",
       "30  pancreatic_endocrinogenesis   \n",
       "5                    noack_2022   \n",
       "14  pancreatic_endocrinogenesis   \n",
       "23  pancreatic_endocrinogenesis   \n",
       "31                         pbmc   \n",
       "38                         pbmc   \n",
       "4                    noack_2022   \n",
       "10                   noack_2022   \n",
       "17  pancreatic_endocrinogenesis   \n",
       "26  pancreatic_endocrinogenesis   \n",
       "34                         pbmc   \n",
       "41                         pbmc   \n",
       "7                    noack_2022   \n",
       "12                   noack_2022   \n",
       "18  pancreatic_endocrinogenesis   \n",
       "27  pancreatic_endocrinogenesis   \n",
       "33                         pbmc   \n",
       "40                         pbmc   \n",
       "1                           gbm   \n",
       "\n",
       "                                                 path  path_exists     n_cells  \n",
       "19  /home/ilibarra/workspace/theislab/mubind/data/...         True         100  \n",
       "28  /home/ilibarra/workspace/theislab/mubind/data/...         True         100  \n",
       "36  /home/ilibarra/workspace/theislab/mubind/data/...         True         100  \n",
       "43  /home/ilibarra/workspace/theislab/mubind/data/...         True         100  \n",
       "6   /home/ilibarra/workspace/theislab/mubind/data/...         True        1000  \n",
       "11  /home/ilibarra/workspace/theislab/mubind/data/...         True        1000  \n",
       "16  /home/ilibarra/workspace/theislab/mubind/data/...         True        1000  \n",
       "25  /home/ilibarra/workspace/theislab/mubind/data/...         True        1000  \n",
       "32  /home/ilibarra/workspace/theislab/mubind/data/...         True        1000  \n",
       "39  /home/ilibarra/workspace/theislab/mubind/data/...         True        1000  \n",
       "2   /home/ilibarra/workspace/theislab/mubind/data/...         True       10000  \n",
       "8   /home/ilibarra/workspace/theislab/mubind/data/...         True       10000  \n",
       "15  /home/ilibarra/workspace/theislab/mubind/data/...         True       10000  \n",
       "24  /home/ilibarra/workspace/theislab/mubind/data/...         True       10000  \n",
       "37  /home/ilibarra/workspace/theislab/mubind/data/...         True       10000  \n",
       "44  /home/ilibarra/workspace/theislab/mubind/data/...         True       10000  \n",
       "20  /home/ilibarra/workspace/theislab/mubind/data/...         True         200  \n",
       "29  /home/ilibarra/workspace/theislab/mubind/data/...         True         200  \n",
       "3   /home/ilibarra/workspace/theislab/mubind/data/...         True        2000  \n",
       "9   /home/ilibarra/workspace/theislab/mubind/data/...         True        2000  \n",
       "0   /home/ilibarra/workspace/theislab/mubind/data/...         True        2000  \n",
       "13  /home/ilibarra/workspace/theislab/mubind/data/...         True        2000  \n",
       "22  /home/ilibarra/workspace/theislab/mubind/data/...         True        2000  \n",
       "35  /home/ilibarra/workspace/theislab/mubind/data/...         True        2000  \n",
       "42  /home/ilibarra/workspace/theislab/mubind/data/...         True        2000  \n",
       "21  /home/ilibarra/workspace/theislab/mubind/data/...         True         300  \n",
       "30  /home/ilibarra/workspace/theislab/mubind/data/...         True         300  \n",
       "5   /home/ilibarra/workspace/theislab/mubind/data/...         True         500  \n",
       "14  /home/ilibarra/workspace/theislab/mubind/data/...         True         500  \n",
       "23  /home/ilibarra/workspace/theislab/mubind/data/...         True         500  \n",
       "31  /home/ilibarra/workspace/theislab/mubind/data/...         True         500  \n",
       "38  /home/ilibarra/workspace/theislab/mubind/data/...         True         500  \n",
       "4   /home/ilibarra/workspace/theislab/mubind/data/...         True        5000  \n",
       "10  /home/ilibarra/workspace/theislab/mubind/data/...         True        5000  \n",
       "17  /home/ilibarra/workspace/theislab/mubind/data/...         True        5000  \n",
       "26  /home/ilibarra/workspace/theislab/mubind/data/...         True        5000  \n",
       "34  /home/ilibarra/workspace/theislab/mubind/data/...         True        5000  \n",
       "41  /home/ilibarra/workspace/theislab/mubind/data/...         True        5000  \n",
       "7   /home/ilibarra/workspace/theislab/mubind/data/...         True         ALL  \n",
       "12  /home/ilibarra/workspace/theislab/mubind/data/...         True         ALL  \n",
       "18  /home/ilibarra/workspace/theislab/mubind/data/...         True         ALL  \n",
       "27  /home/ilibarra/workspace/theislab/mubind/data/...         True         ALL  \n",
       "33  /home/ilibarra/workspace/theislab/mubind/data/...         True         ALL  \n",
       "40  /home/ilibarra/workspace/theislab/mubind/data/...         True         ALL  \n",
       "1   /home/ilibarra/workspace/theislab/mubind/data/...         True  compressed  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import anndata\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "# genomes_path = '/mnt/c/Users/IgnacioIbarra/Dropbox/annotations/%s/genome/%s.fa'\n",
    "genomes_path = '/home/ilibarra/annotations/%s/genome/%s.fa'\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "df = pd.DataFrame(ad_paths, columns=['dataset', 'path'])\n",
    "df['path_exists'] = [os.path.exists(p) for p in df['path']]\n",
    "obs = [f.split('_')[-2] for f in list(df['path'].str.split('/').str[-1])]\n",
    "obs = [o.replace('obs', '') for o in obs]\n",
    "df['n_cells'] = obs\n",
    "df.sort_values(['n_cells', 'dataset'], ascending=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(45, 4)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# custom query (500 cells)\n",
    "sel_df = df # df[df['n_cells'] == '500']\n",
    "sel_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "overwrite_prepare = False\n",
    "overwrite_train = False\n",
    "query_epochs = [50,] #  10, 20] # 50] #  100] #  ,50, 100] #   10, 20, 50, 100] #  5, 10, 50, 100]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_missing = 0\n",
    "n_folds = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'query_epochs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m n_epochs \u001b[38;5;129;01min\u001b[39;00m \u001b[43mquery_epochs\u001b[49m:\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m ri, r \u001b[38;5;129;01min\u001b[39;00m sel_df[sel_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mn_cells\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m1000\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39miterrows(): \u001b[38;5;66;03m# [sel_df['n_cells'].isin({'100', '200', '300'})].iterrows():\u001b[39;00m\n\u001b[1;32m      3\u001b[0m         k, ad_path \u001b[38;5;241m=\u001b[39m r[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdataset\u001b[39m\u001b[38;5;124m'\u001b[39m], r[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpath\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'query_epochs' is not defined"
     ]
    }
   ],
   "source": [
    "for n_epochs in query_epochs:\n",
    "    for ri, r in sel_df[sel_df['n_cells'] == '1000'].iterrows(): # [sel_df['n_cells'].isin({'100', '200', '300'})].iterrows():\n",
    "        k, ad_path = r['dataset'], r['path']\n",
    "\n",
    "        print(r.values)\n",
    "\n",
    "        # if k != 'pbmc':\n",
    "        #     continue\n",
    "\n",
    "        print('here...')\n",
    "        # try:\n",
    "        # print(k)\n",
    "        # if not 'noack_2022' in k:\n",
    "        #     continue\n",
    "        # if 'ALL' in ad_path: #  and not '_obs500_' in ad_path: #  or not '/random/' in ad_path:\n",
    "        #     continue\n",
    "        # print(os.path.exists(ad_path), ad_path)\n",
    "        # print('\\nnext path')\n",
    "        print(k, ad_path)\n",
    "        n_cells = ad_path.split('_')[-2]\n",
    "        sampling_method = ad_path.split('/')[-2]\n",
    "\n",
    "        # print(k, sampling_method, ad_path)\n",
    "\n",
    "        species = 'hg38' if (k != 'noack_2022' and k != 'pancreatic_endocrinogenesis') else 'mm10'\n",
    "\n",
    "        filename = os.path.basename(ad_path)\n",
    "\n",
    "\n",
    "        for fold_i in range(n_folds):\n",
    "            input_dir = os.path.join(ad_path.replace(filename, 'scbasset_input'),\n",
    "                                     n_cells,\n",
    "                                     'fold_%i' % fold_i)\n",
    "\n",
    "            print('')\n",
    "            print('# fold %i' % fold_i)\n",
    "            for use_poisson in [1, 0]:\n",
    "                out_dir = os.path.join(ad_path.replace(filename,\n",
    "                                                    'scbasset_output'),\n",
    "                                                    n_cells + '_e%i' % n_epochs,\n",
    "                                                    'poisson' if use_poisson else 'bce',\n",
    "                                                    'fold_%i' % fold_i)\n",
    "                \n",
    "                # print('in', input_dir)\n",
    "                # print('out', os.path.exists(out_dir), out_dir)\n",
    "                # assert False\n",
    "\n",
    "                if not os.path.exists(input_dir):\n",
    "                    os.makedirs(input_dir)\n",
    "                if not os.path.exists(out_dir):\n",
    "                    os.makedirs(out_dir)\n",
    "                \n",
    "                # print(species)\n",
    "                train_path = os.path.join(input_dir, 'fold_%i' % fold_i, 'ad.h5ad')\n",
    "                train_seqs_path = os.path.join(input_dir, 'fold_%i' % fold_i, 'train_seqs.h5')\n",
    "\n",
    "                print(os.path.exists(train_path), train_path)\n",
    "                print(os.path.exists(train_seqs_path), train_seqs_path)\n",
    "                # assert False\n",
    "\n",
    "                outpath_time = os.path.join(out_dir, 'running_time.pkl')\n",
    "\n",
    "                time_ok = False\n",
    "                if os.path.exists(outpath_time):\n",
    "                    running_time = pickle.load(open(outpath_time, 'rb'))\n",
    "                    time_ok = isinstance(running_time, datetime.timedelta)\n",
    "                    # print('time ok?', time_ok, running_time)\n",
    "\n",
    "                print('here...')\n",
    "                \n",
    "                print(k, not os.path.exists(train_path) or not os.path.exists(train_seqs_path) or overwrite_prepare)\n",
    "                if not os.path.exists(train_path) or not os.path.exists(train_seqs_path) or overwrite_prepare:\n",
    "                    # print('preparing data using path....')\n",
    "                    # print(ad_path)\n",
    "                    adata = anndata.read_h5ad(ad_path)\n",
    "                    # assert False\n",
    "\n",
    "                    print(adata.shape)\n",
    "                    print('parsing...')\n",
    "\n",
    "                    # print(adata.var[['chr', 'start', 'end']])\n",
    "                    adata.var['chr'] = np.where(adata.var_names.str.contains(':'), adata.var_names.str.split(':').str[0], adata.var_names.str.split('-').str[0])\n",
    "                    adata.var['chr'] = np.where(~adata.var['chr'].str.contains('chr'), 'chr', '') + adata.var['chr']\n",
    "                    # adata.var['start'] = np.where(~adata.var_names.str.contains(':'), adata.var_names.str.split('-').str[1], adata.var_names.str.split(':').str[1].str.split('-').str[0])\n",
    "                    # adata.var['end'] = np.where(~adata.var_names.str.contains(':'), adata.var_names.str.split('-').str[2], adata.var_names.str.split('-').str[1])\n",
    "\n",
    "                    colon_found = adata.var_names.str.contains(':').any()\n",
    "                    colon_found\n",
    "                    if colon_found:\n",
    "                        adata.var['start'] =   adata.var_names.str.split(':').str[1].str.split('-').str[0]\n",
    "                        adata.var['end'] = adata.var_names.str.split(':').str[1].str.split('-').str[1]\n",
    "                    else:\n",
    "                        adata.var['start'] = adata.var_names.str.split('-').str[1]\n",
    "                        adata.var['end'] = adata.var_names.str.split('-').str[2]\n",
    "                        \n",
    "                    # remove non-canonical chromosomes\n",
    "\n",
    "                    adata = adata[:,~adata.var['chr'].str.contains('chrGL') & ~adata.var['chr'].str.contains('chrK')].copy()\n",
    "                    adata.write(ad_path, compression='lzf')\n",
    "                    genome_fasta = genomes_path % (species, species)\n",
    "                    # os.path.exists(genome_fasta)\n",
    "                    # print(genome_fasta)\n",
    "                    # preprocess fasta\n",
    "                    print('preprocess fasta')\n",
    "\n",
    "                    # batch = 5 assures preparing a small batch\n",
    "                    !echo $pythonbin scbasset_preprocess.py --ad_file $ad_path --input_fasta $genome_fasta --out_path $input_dir --batch 5 --seed $fold_i\n",
    "                    !$pythonbin scbasset_preprocess.py --ad_file $ad_path --input_fasta $genome_fasta --out_path $input_dir --batch 5 --seed $fold_i\n",
    "                    # print('')\n",
    "                else:\n",
    "                    # continue\n",
    "                    print('skip prepare (already done...)')\n",
    "\n",
    "                # assert False\n",
    "                # continue\n",
    "\n",
    "                best_model_path = os.path.join(out_dir, 'best_model.h5')\n",
    "                print(os.path.exists(best_model_path), best_model_path)\n",
    "                if not os.path.exists(best_model_path) or overwrite_train or not os.path.exists(outpath_time) or not time_ok: \n",
    "                    # print('')\n",
    "                    print(ad_path)\n",
    "\n",
    "                    !echo $pythonbin scbasset_train.py --input_folder $input_dir --epochs $n_epochs --out_path $out_dir\n",
    "                    print('about to train...')\n",
    "\n",
    "                    t1 = datetime.datetime.now()\n",
    "\n",
    "                    n_missing += 1\n",
    "                    # continue\n",
    "                    # assert False\n",
    "                    !$pythonbin scbasset_train.py --input_folder $input_dir --epochs $n_epochs --out_path $out_dir --use_poisson $use_poisson\n",
    "                    print('')\n",
    "                    t2 = datetime.datetime.now()\n",
    "                    time_diff = (t2 - t1)\n",
    "\n",
    "                    pickle.dump(time_diff, open(outpath_time, 'wb'))\n",
    "                    print(os.path.exists(outpath_time), outpath_time)\n",
    "                    # assert False\n",
    "                else:\n",
    "                    print('error...')\n",
    "                    continue\n",
    "                    print('skip train (already done...)')\n",
    "                # print('')\n",
    "            # except Exception:\n",
    "            #     pass\n",
    "\n",
    "        # print('')\n",
    "\n",
    "print('n_missing = %i' % n_missing)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mubind",
   "language": "python",
   "name": "mubind"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
